### PR TITLE
[YUNIKORN-1435] Expose RM diagnostics in full state dump

### PR DESCRIPTION
**Pull request by craigcondit**
### What is this PR for?
Register the RM as a `StateDumpPlugin` so that it can participate in state dumps. During a full state dump, the contents of the scheduler cache and internal data structures will be output.

NOTE: This PR will be in draft status until https://github.com/apache/yunikorn-core/pull/459 is merged, as the Core / SI must be updated first.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1435

### How should this be tested?
Unit tests have been added to verify basic functionality.

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by craigcondit** [comment]
Note for reviewers: The massive go.mod / go.sum changes were needed as something broke the build upstream (I suspect the prometheus update).

**Review by manirajv06 — CHANGES_REQUESTED** [review]
changes looks good. overall test coverage needs to be improved lot specifically for each object dao inside dump.

File: pkg/cache/external/scheduler_cache_dao.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+type ContainerDao struct {
+	Name      string                  `json:"name"`
+	Resources v1.ResourceRequirements `json:"resources"`
+}
+
+type PodSchedulingInfoDao struct {
+	Namespace       string    `json:"-"`
+	Name            string    `json:"-"`
```
Comment (manirajv06) [inline pkg/cache/external/scheduler_cache_dao.go:87]: Is "-" intentional?

**Review by manirajv06 — CHANGES_REQUESTED** [review]

File: pkg/cache/external/scheduler_cache_dao.go (hunk 1 above)
Comment (craigcondit) [reply pkg/cache/external/scheduler_cache_dao.go:87]: Yes, we don't want these to emit via JS.

**Comment by craigcondit** [comment]
> changes looks good. overall test coverage needs to be improved lot specifically for each object dao inside dump.

The objects themselves are simply copies of K8s objects. The new test cases verify that they are created correctly.

**Comment by manirajv06** [comment]
> The objects themselves are simply copies of K8s objects.

Not all, I guess. For ex, assignedPods, assumedPods, pendingAllocations etc. Even I am not going by code coverage tool results. Yes, objects creation are asserted. I meant the assertion to verify the object properties/details. May be, we can file a newbie jira to cover these properties in detail.

**Review by manirajv06 — APPROVED** [review]

**Pull request closed without merging**